### PR TITLE
[lightspark-sdk] Add integration tests

### PR DESCRIPTION
**Pull request by Sorizen**

### CODE REVIEW

File: packages/lightspark-sdk/package.json [hunk 1]
```diff
@@ -82,7 +82,6 @@
     "postversion": "yarn build",
     "test-cmd": "node --experimental-vm-modules $(yarn bin jest) --no-cache --runInBand --bail",
     "test": "yarn test-cmd src/tests/*.test.ts",
-    "test:integration": "yarn test-cmd src/tests/integration/*.test.ts",
```
Comment (coreymartin) [inline packages/lightspark-sdk/package.json:85]: These need to be separate commands.

File: packages/lightspark-sdk/src/tests/general-regtest.test.ts [hunk 2]
```diff
@@ -0,0 +1,462 @@
+import LightsparkClient from '../client.js'
```
Comment (coreymartin) [inline packages/lightspark-sdk/src/tests/general-regtest.test.ts:1]: Move to integration folder

File: packages/lightspark-sdk/src/tests/general-mainnet.test.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+describe('test', () => {
```
Comment (coreymartin) [inline packages/lightspark-sdk/src/tests/general-mainnet.test.ts:1]: Move to integration folder

File: packages/lightspark-sdk/src/tests/general-regtest.test.ts [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+        expect(regtestNode).toBeDefined()
+    }, TESTS_TIMEOUT)
+})
+
+describe('P0 tests', () => {
+
+    test('Should create a normal payment invoice', async () => {
+        if(!regtestNodeId) {
+            throw new TypeError("No regtest nodes in account")
+        }
```
Comment (coreymartin) [inline packages/lightspark-sdk/src/tests/general-regtest.test.ts:84]: For all regtestNodeId checks use [this function](https://github.com/lightsparkdev/js-sdk/blob/main/packages/lightspark-sdk/src/tests/integration/client.test.ts#L42-L48) instead. This asserts that it exists and guarantees the type returned to not be undefined. More concise and more importantly makes it super clear why the tests are failing.

File: packages/lightspark-sdk/src/tests/general-regtest.test.ts [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+    OutgoingPayment,
+    TransactionStatus
+} from '../index.js'
+
+const { apiTokenClientId, apiTokenClientSecret, baseUrl } =
+    getCredentialsFromEnvOrThrow()
+
+let regtestClient: LightsparkClient
+
+const unauthorizedRegtestClient = new LightsparkClient()
```
Comment (coreymartin) [inline packages/lightspark-sdk/src/tests/general-regtest.test.ts:31]: There's nothing about these clients that makes them only capable of REGTEST operations. Please rename to lightsparkClient to make that clear.

File: packages/lightspark-sdk/package.json (hunk 1 above)
Comment (Sorizen) [reply packages/lightspark-sdk/package.json:85]: done

File: packages/lightspark-sdk/src/tests/general-regtest.test.ts (hunk 2 above)
Comment (Sorizen) [reply packages/lightspark-sdk/src/tests/general-regtest.test.ts:9]: done

File: packages/lightspark-sdk/src/tests/general-mainnet.test.ts (hunk 3 above)
Comment (Sorizen) [reply packages/lightspark-sdk/src/tests/general-mainnet.test.ts:1]: done

File: packages/lightspark-sdk/src/tests/general-regtest.test.ts (hunk 5 above)
Comment (Sorizen) [reply packages/lightspark-sdk/src/tests/general-regtest.test.ts]: done

File: packages/lightspark-sdk/src/tests/general-regtest.test.ts (hunk 4 above)
Comment (Sorizen) [reply packages/lightspark-sdk/src/tests/general-regtest.test.ts]: done

File: .yarn/releases/yarn-3.6.4.cjs
Comment (coreymartin) [inline .yarn/releases/yarn-3.6.4.cjs:1]: This file shouldn't be modified, mind reverting it?

Comment (coreymartin) [reply .yarn/releases/yarn-3.6.4.cjs:1]: To get the exact version of yarn installed based on project settings follow [these instructions](https://github.com/lightsparkdev/js-sdk#running-the-sdks-and-examples)

File: .yarnrc.yml [hunk 6]
```diff
@@ -1,2 +1,3 @@
-# Required for react-native support:
 nodeLinker: node-modules
+
+yarnPath: .yarn/releases/yarn-3.6.4.cjs
```
Comment (coreymartin) [inline .yarnrc.yml:3]: Revert these changes please

File: packages/lightspark-sdk/package.json [hunk 7]
```diff
@@ -93,7 +93,7 @@
     "auto-bind": "^5.0.1",
     "crypto": "^1.0.1",
     "crypto-browserify": "^3.12.0",
-    "dayjs": "^1.11.7",
+    "dayjs": "^1.11.10",
```
Comment (coreymartin) [inline packages/lightspark-sdk/package.json:96]: Needs to be the 1.11.7 otherwise workspace checks won't pass

File: packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+                publicKey: b64encode(
+                    await DefaultCrypto.serializeSigningKey(publicKey, 'spki'),
+                ),
+            }
+
+            expect(serializedKeypair.privateKey).not.toBeNull()
+            expect(serializedKeypair.publicKey).not.toBeNull()
+        },
+        TESTS_TIMEOUT,
+    )
```
Comment (coreymartin) [inline packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts:226]: We should move this to core tests (core/src/crypto/crypto.test.ts) since it's not testing any lightspark-sdk function. Testing setup there should work the exact same so it should just be copy/paste.

I think for "key generation" that @jklein24 has listed in P1 tests he may be thinking of [these commands in lightspark-cli](https://github.com/lightsparkdev/js-sdk/blob/main/packages/lightspark-cli/src/index.ts#L494-L556). We could move those functions into lightspark-sdk so that they're testable here, @jklein24 wdyt?

File: packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+
+    test(
+        'should listen current payment requests',
+        async () => {
+            const requests = await lightsparkClient.getRecentPaymentRequests(
+                getRegtestNodeId(),
+                PAGINATION_STEP,
+                BitcoinNetwork.REGTEST,
+            )
+            expect(requests.length).toBe(PAGINATION_STEP)
```
Comment (coreymartin) [inline packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts:253]:
```suggestion
            // There will be more transactions than paginated due to repeat testing on this account, so the amount returned should equal PAGINATION_STEP
            expect(requests.length).toBe(PAGINATION_STEP)
```
Let's add this comment so it's clear why transactions will equal PAGINATION_STEP

File: packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+        'should listen current payment requests after some date',
+        async () => {
+            const requestsAfterDate = day(Date.now() - DAY_IN_MS).format()
+            const requests = await lightsparkClient.getRecentPaymentRequests(
+                getRegtestNodeId(),
+                PAGINATION_STEP,
+                BitcoinNetwork.REGTEST,
+                requestsAfterDate
+            )
+            expect(requests.length).toBe(PAGINATION_STEP)
```
Comment (coreymartin) [inline packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts:268]: How do we know for sure that there will be PAGINATION_STEP payment requests in the last day?

File: packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts (hunk 8 above)
Comment (jklein24) [reply packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts:226]: Ah yeah, that's a good point... I suppose key generation is slightly less critical for lightspark-sdk than for wallet-sdk since it only happens once and can be out-of-band from the actual code. Agreed that moving this to core tests makes sense to me.

File: .yarn/releases/yarn-3.6.4.cjs
Comment (Sorizen) [reply .yarn/releases/yarn-3.6.4.cjs:1]: reverted

File: .yarnrc.yml (hunk 6 above)
Comment (Sorizen) [reply .yarnrc.yml]: reverted

File: packages/lightspark-sdk/package.json (hunk 7 above)
Comment (Sorizen) [reply packages/lightspark-sdk/package.json]: done

File: packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts (hunk 8 above)
Comment (Sorizen) [reply packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts]: Added to core

File: packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts (hunk 9 above)
Comment (Sorizen) [reply packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts:239]: Created a function that will create 20 transactions before this tests

Comment (coreymartin) [reply packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts:239]: We should do less than 20 to save time and prevent flakiness. How about 3?

Comment (Sorizen) [reply packages/lightspark-sdk/src/tests/integration/general-regtest.test.ts:239]: I just reduced the pagination step from 30 to 3. I think this will solve the issue.
